lua/lean/_util.lua

```diff
@@ -50,4 +50,48 @@ function M.subprocess_check_output(opts, timeout)
   ))
 end
 
+function M.after_or_equal(pos, other_pos)
+  if pos.line > other_pos.line then return true end
+  if pos.line == other_pos.line and pos.character >= other_pos.character then return true end
+  return false
+end
+
+function M.update_position(pos, changes)
+  local new_pos = vim.deepcopy(pos)
+
+  for _, change in pairs(changes) do
+    local start_pos = change.range["start"]
+    local end_pos = change.range["end"]
+    if not M.after_or_equal(new_pos, start_pos) then goto next_change end
+
+    local new_lines = vim.split(change.text, "\n")
+    if new_pos.line >= end_pos.line then
+      local orig_line_offset = new_pos.line - end_pos.line
+      local new_end_line = start_pos.line + (#new_lines - 1)
+      new_pos.line = new_end_line + orig_line_offset
+
+      if new_pos.line == end_pos.line then
+        -- change range is exclusive, so okay if ==
+        if new_pos.character >= end_pos.character then
+          local orig_char_offset = new_pos.character - end_pos.character
+          local new_end_char = #new_lines == 1 and (start_pos.character + #new_lines[1])
+            or #(new_lines[#new_lines])
+          new_pos.character = new_end_char + orig_char_offset
+        else
+          -- within modified range, invalidate
+          new_pos = nil
+          return new_pos
+        end
+      end
+    else
+      -- within modified range, invalidate
+      new_pos = nil
+      return new_pos
+    end
+    ::next_change::
+  end
+
+  return new_pos
+end
```

Comment:
@gebner The logic for updating pin positions in VSCode [seemed to be lacking a bit](https://github.com/leanprover-community/vscode-lean4/blob/b7382e3b684baa2c211cf8f4d8e7f1681f99311c/lean4-infoview/src/infoview/infos.tsx#L40), this seemed much more accurate, but let me know if there's a particular reason it was done that way which I'm overlooking.

Reply:
The reason why the vscode extension parses the LSP notification is because Wojciech wanted to make the infoview as editor-independent as possible and vscode doesn't have an API to track document positions after changes.  (So you'd either parse the changes from the LSP event or from the vscode event.)

Is there a reason we can't use extmarks in neovim?

Reply:
Yep we're doing the same here, the parameters used by this method come directly from `textDocument/didChange`. I thought this would be preferable to something built-in because it guarantees that pin positions are kept exactly in sync with the server's version of the text.

In particular if we are debouncing changes (have yet to write tests for this) if the pin positions are updated before sending `textDocument/didChange` and `$/lean/fileProgress` triggers a request, it could use positions that are invalid from the server's perspective.

Having said that, extmarks are new to me so yes I may have reinvented a bit here, will look into whether we can use this.

Reply:
> if the pin positions are updated before sending textDocument/didChange

This is indeed a problem that is avoided by tracking the LSP change notifications.  But I'm not sure how much we gain by reimplementing extmarks just for pins alone.  All other commands (hover, autocomplete, etc.) that we provide assume that the document is in sync.

Reply:
Done (removed the custom math here in favor of extmarks). But I still think we should have some extra assurance about the document being in sync for updating the infoview (which we do have right now), since we'll want that to be able to happen at any time (e.g. on `$/lean/fileProgress`). So it's a bit different from the other commands you mentioned , since they're more directly interactive and so you have more of a guarantee of things being in sync (particularly if they happen after a timer).